plan/rule_aggregation_elimination.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+)
+
+type aggregationEliminater struct {
+}
+
+// tryToEliminateAggregation will eliminate aggregation grouped by unique key.
+// e.g. select min(b) from t group by a. If a is a unique key, then this sql is equal to `select b from t group by a`.
+// For count(expr), sum(expr), avg(expr), count(distinct expr, [expr...]) we may need to rewrite the expr. Details are shown below.
+// If we can eliminate agg successful, we return a projection. Else we return a nil pointer.
+func (a *aggregationEliminater) tryToEliminateAggregation(agg *LogicalAggregation) *LogicalProjection {
+	schemaByGroupby := expression.NewSchema(agg.groupByCols...)
+	coveredByUniqueKey := false
+	for _, key := range agg.children[0].Schema().Keys {
+		if schemaByGroupby.ColumnsIndices(key) != nil {
+			coveredByUniqueKey = true
+			break
+		}
+	}
+	if coveredByUniqueKey {
+		// GroupByCols has unique key, so this aggregation can be removed.
+		proj := a.convertAggToProj(agg)
+		proj.SetChildren(agg.children[0])
+		return proj
+	}
+	return nil
+}
+
+func (a *aggregationEliminater) convertAggToProj(agg *LogicalAggregation) *LogicalProjection {
```

Comment:
Most methods are copied from `aggregate push down`. I think it's not good to leave them as function that anyone can call.
Extract a struct that can be used both by `aggregate elimination` and `aggregate push down`, make them as method of this struct ?

Reply:
Hmm, can we set `aggregationOptimizer` as an anonymous field of `aggregationEliminater`? then we only need to override the `optimize` method for `aggregationEliminater`.

`aggregationOptimizer` also handles the aggregate elimination, why not use it here? because `AllowAggPushDown` is false by default?

> Extract a struct that can be used both by aggregate elimination and aggregate push down, make them as method of this struct ?

IMHO, `aggregationOptimizer` should be this struct, and we should add an `aggregationPushDownSolver` as subclass of it to hold the aggregate pushdown logic.

Reply:
> `aggregationOptimizer` also handles the aggregate elimination, why not use it here? because `AllowAggPushDown` is false by default?
Yes, that one is mainly to push down aggregation. The elimination is done there just for convenience.

> IMHO, `aggregationOptimizer` should be this struct, and we should add an `aggregationPushDownSolver` as subclass of it to hold the aggregate pushdown logic.
But there's no overriding in golang. So we cannot make one containing another one. They both have a recursive method `optimize` to implement `logicalOptRule` interface.